agent/dns.go

```diff
@@ -1099,50 +1116,47 @@ func trimTCPResponse(req, resp *dns.Msg) (trimmed bool) {
 	if len(resp.Answer) > truncateAt {
 		resp.Answer = resp.Answer[:truncateAt]
 	}
-	if hasExtra {
-		index = make(map[string]dns.RR, len(resp.Extra))
-		indexRRs(resp.Extra, index)
-	}
-	truncated := false
-
-	// This enforces the given limit on 64k, the max limit for DNS messages
-	for len(resp.Answer) > 1 && resp.Len() > maxSize {
-		truncated = true
-		// first try to remove the NS section may be it will truncate enough
-		if len(resp.Ns) != 0 {
-			resp.Ns = []dns.RR{}
-		}
-		// More than 100 bytes, find with a binary search
-		if resp.Len()-maxSize > 100 {
-			bestIndex := dnsBinaryTruncate(resp, maxSize, index, hasExtra)
-			resp.Answer = resp.Answer[:bestIndex]
-		} else {
-			resp.Answer = resp.Answer[:len(resp.Answer)-1]
-		}
-		if hasExtra {
-			syncExtra(index, resp)
-		}
-	}
 
-	return truncated
+	return trimResponseGeneric(req, resp, maxSize)
 }
 
-// trimUDPResponse makes sure a UDP response is not longer than allowed by RFC
-// 1035. Enforce an arbitrary limit that can be further ratcheted down by
-// config, and then make sure the response doesn't exceed 512 bytes. Any extra
-// records will be trimmed along with answers.
-func trimUDPResponse(req, resp *dns.Msg, udpAnswerLimit int) (trimmed bool) {
-	numAnswers := len(resp.Answer)
-	hasExtra := len(resp.Extra) > 0
-	maxSize := defaultMaxUDPSize
+// trimUDPResponse makes sure a UDP response is not longer than allowed: 512 bytes
+// without EDNS (per RFC 1035), or a user specified limit with EDNS. Any Extra
+// records will be trimmed along with Answers.
+func trimUDPResponse(req, resp *dns.Msg) (trimmed bool) {
+	maxSizeTotal := defaultMaxUDPSize
 
 	// Update to the maximum edns size
 	if edns := req.IsEdns0(); edns != nil {
-		if size := edns.UDPSize(); size > uint16(maxSize) {
-			maxSize = int(size)
+		if size := edns.UDPSize(); size > uint16(maxSizeTotal) {
+			maxSizeTotal = int(size)
 		}
 	}
 
+	// NOTE: 7 is a magic number added in commit 4f5d502 to stop an "overflow
```

Comment:
Note to reviewer: I did not attempt to fix the magic numbers, but I did want to describe them so that someone with more DNS domain knowledge would have an easier time fixing these numbers in the future.